File: R/mmMCMCFit.R
#' Fit Mixed Membership models using a Metropolis Hastings within Gibbs MCMC sampler
#' 
#' Takes samples form the posterior distribution of a mixed membership model using a 
#' Metropolis-Hastings within Gibbs sampler. 
#' 
#' \code{mmMCMCFit} draws samples from the posterior distribution of a mixed membership model
#' given a set of observed data. The sampler takes Metropolis Hastings steps to sample the \eqn{\alpha_0}
#' and \eqn{\xi} parameters and samples the remaining \eqn{\theta}, \eqn{\lambda_i} and \eqn{Z} parameters
#' with a Gibbs sampler.   
#' 
#' @param model a \code{mixedMemModelMCMC} object created by the \code{mixedMemModelMCMC} constructor
#' @param burnIn non-negative integer indicating the number of burn in samples before recording the first sample
#' @param samples positive integer indicating the number of samples to record
#' @param thin positive integer indicating how to thin the samples
#' @param print positive integer indicating how often to print an update to the R console
#' @param fileNames list of files names to write samples
#' @param newFiles 0 if samples should be appended to existing files; 1 if samples should overwrite any existing files 
#' @param omega tuning parameter for MH step for alpha.
#' @param eta tuning parameter for MH step for ksi
#' @param whichWrite which parameters to write to disk. Writing the individual parameters can significantly increase computational
#' time when the number of individuals is large
#' @param extended boolean whether to estimate the extended model or not
#' @seealso mixedMemModelMCMC
#' @export
mmMCMCFit <- function(model, burnIn = 20000, samples = 1000, thin = 10, print = 100, fileNames, newFiles = 1,
                      omega = 100, eta = 1, whichWrite = c(1, 1, 1, 0, 0, 1, 0)) {
  
  # check to make sure the model is of the right class
  if(class(model) != "mixedMemModelMCMC"){
    stop("Input model must be of class mixedMemModelMCMC")
  }
  
  # checks inputs for internal consistency
  checkModelMCMC(model) 
  
  # In C code, Z must be 0:(K-1), but in R it is it 1:K so we modify it before inputting to the c++ function
  model$Z <- model$Z -1
  mcmcInputC(model = model, burnIn = burnIn, samples = samples, thin = thin, print = print, fileNames = fileNames,
                      newFiles = newFiles, omega = omega, eta = eta, whichWrite = whichWrite)
  
  
}

File: src/mmMCMCFit.cpp
#include "mmMCMCFit.h"

void mixedMemMcmcC(mm_model_mcmc model, int burnIn, int samples, int thin, int print, Rcpp::List fileNames, int newFiles,
                   double omega, double eta, NumericVector which_write)
{
    if(newFiles) {
        clearFiles(fileNames, which_write);
    }

    int t,s;

    for(t = 0; t < burnIn; t++) {
        if(t% print == 0) {
            Rcout<<"Burn-in: "<<t<<std::endl;
        }

        //Data Augmentation
        if(model.getExtended() == 1) {
            imputeStayers(model);
        }

        imputeLatent(model);


        // posterior draws of model parameters
        sampleTheta(model) ;

        sampleKsi(model, eta);
        // Draw dirichlet parameter
        sampleAlpha(model, omega);


        // Draw Extended
        if(model.getExtended() == 1) {
            sampleP(model);
        }
    }

    Rcout<<"=== Done with Burn In ==="<<std::endl;
    for(t = 0; t < samples; t++) {
        for(s = 0; s < thin; s++) {

            //Data Augmentation
            if(model.getExtended() == 1) {
                imputeStayers(model);
            }
            imputeLatent(model);

            // posterior draws of model parameters
            sampleTheta(model);

            sampleKsi(model, eta);

            // Draw dirichlet parameter
            sampleAlpha(model, omega);

            // Sample P
            if(model.getExtended() == 1) {
                sampleP(model);
            }

        }

        if(t% print == 0) {
            Rcout<<"Sample : "<<t<<std::endl;
        }

        writeState(model, fileNames, which_write);
    }

    Rcout<<"Sampling Completed!"<<std::endl;
}


void writeState(mm_model_mcmc model, Rcpp::List fileNames, NumericVector which_write)
{
    int i, j, k, r, v;
    std::ofstream myfile;
    std::string file;


    //Write Theta
    if(which_write[0] == 1) {
        file = as<std::string> (fileNames[0]);
        myfile.open (file.c_str(),std::ios::app);
        for(v = 0; v < model.getMaxV(); v++) {
            for(k = 0; k < model.getK(); k++) {
                for(j = 0; j < model.getJ(); j++) {
                    if((j + k + v) > 0) {
                        myfile<< ", ";
                    }
                    if(v < model.getV(j)) {
                        myfile << model.getTheta(j, k, v);
                    } else {
                        myfile << 0;
                    }
                }
            }
        }
        myfile<<std::endl;
        myfile.close();
    }

// Write Alpha
    if(which_write[1] == 1) {
        file = as<std::string> (fileNames[1]);
        myfile.open (file.c_str(),std::ios::app);
        myfile << model.getAlpha();
        myfile << std::endl;
        myfile.close();
    }

    //Write ksi
    if(which_write[2] == 1) {
        //Write ksi
        file = as<std::string> (fileNames[2]);
        myfile.open(file.c_str(),std::ios::app);
        for(k = 0; k <model.getK(); k++) {
            if(k > 0) {
                myfile<< ", ";
            }
            myfile<< model.getKsi(k);
        }
        myfile<<std::endl;
        myfile.close();

    }


    //Write Lambda
    if(which_write[3] == 1) {
        file = as<std::string> (fileNames[3]);
        myfile.open (file.c_str() ,std::ios::app);
        for(k = 0; k < model.getK(); k++) {
            for(i = 0; i < model.getT(); i++) {
                if((i+k) > 0) {
                    myfile<< ", ";
                }
                myfile << model.getLambda(i,k);
            }
        }
        myfile << std::endl;
        myfile.close();
    }

    //write Z
    if(which_write[4] == 1) {

        file = as<std::string> (fileNames[4]);
        myfile.open (file.c_str(),std::ios::app);
        for(r = 0; r < model.getMaxR(); r++) {
            for(j = 0; j < model.getJ(); j++) {
                for(i = 0; i < model.getT(); i++) {
                    if((i + j + r) > 0) {
                        myfile<< ", ";
                    }
                    if(r < model.getR(j)) {
                        myfile << model.getZ(i,j,r);
                    } else {
                        myfile << -1 ;
                    }
                }
            }
        }
        myfile<<std::endl;
        myfile.close();

    }


    // Write P
    if(which_write[5] == 1) {
        file = as<std::string> (fileNames[5]);
        myfile.open (file.c_str(),std::ios::app);
        for(k = 0; k < (model.getS() + 1); k++) {
            if(k > 0) {
                myfile<< ", ";
            }
            myfile << model.getP(k);
        }
        myfile<<std::endl;
        myfile.close();
    }

    // Write stayerStatus
    if(which_write[6] == 1) {
        file = as<std::string> (fileNames[6]);
        myfile.open (file.c_str(),std::ios::app);
        for(i = 0; i < model.getT() ; i ++) {
            if(i > 0) {
                myfile << ", ";
            }
            myfile << model.getStayerStatus(i);
        }
        myfile<<std::endl;
        myfile.close();

    }
}

void clearFiles(Rcpp::List fileNames, NumericVector which_write)
{
    std::ofstream myfile;
    std::string file;


    if(which_write[0] == 1) {
        file = as<std::string> (fileNames[0]);
        myfile.open(file.c_str());
        myfile.close();
    }

    if(which_write[1] == 1) {
        file = as<std::string> (fileNames[1]);
        myfile.open(file.c_str());
        myfile.close();
    }

    if(which_write[2] == 1) {
        file = as<std::string> (fileNames[2]);
        myfile.open(file.c_str());
        myfile.close();
    }

    if(which_write[3] == 1) {
        file = as<std::string> (fileNames[3]);
        myfile.open(file.c_str());
        myfile.close();
    }

    if(which_write[4] == 1) {
        file = as<std::string> (fileNames[4]);
        myfile.open(file.c_str());
        myfile.close();
    }

    if(which_write[5] == 1) {
        file = as<std::string> (fileNames[5]);
        myfile.open(file.c_str());
        myfile.close();
    }

    if(which_write[6] == 1) {
        file = as<std::string> (fileNames[6]);
        myfile.open(file.c_str());
        myfile.close();
    }
}
